---
title: "Shed Level Early Infection - Bursa - Gene"
author: "Amanda Dolinski & Jared J. Homola"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  pdf_document: default
  html_document: default
  word_document: default
---

\pagenumbering{gobble}

This is an analysis of differential expression at the gene level between low, moderate, and high LPAIV shedding blue-winged teals based on cloacal swab virus titers measured on the day of sacrifice for bursa samples on 1 and 3 days post infection (DPI).

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, fig.pos = 'H')
```

```{r message=FALSE, warning=FALSE}
## Load packages and data
library(limma)
library(edgeR)
library(Glimma)
library(gplots)
library(ggrepel)
library(RColorBrewer)
library(clusterProfiler)
library(kableExtra)
library(gridExtra)
library(topGO)
library(tidyverse)

## Load data
annot <- read.delim("G:/Shared drives/RNA Seq Supershedder Project/BWTE DE manuscript/extData/Trinotate.csv", header = TRUE, sep = "\t")
cnt <- read.table("G:/Shared drives/RNA Seq Supershedder Project/BWTE DE manuscript/extData/rsem.gene.counts.matrix", header = TRUE)
covars <- read.csv("G:/Shared drives/RNA Seq Supershedder Project/BWTE DE manuscript/extData/BWTE54_SSgroup_Raw_Pool.csv", header = TRUE)
```

```{r, warning = FALSE, message = FALSE}
cnt <- cnt %>%
  select(
    -alignEstimateAbundance_BWTE_Ileum_36_S50,
    -alignEstimateAbundance_BWTE_Bursa_36_S31,-alignEstimateAbundance_BWTE_Ileum_19_S35,
    -alignEstimateAbundance_BWTE_Bursa_19_S14) %>%
  rownames_to_column("gene") %>%
  separate(gene, into = c(NA, "pt1", "pt2", "gene")) %>%
  unite(gene, pt1, pt2, gene) %>%
  column_to_rownames("gene") %>% 
  select(contains("Bursa"))

covars <- covars %>%
  filter(!bird %in% c("36", "19")) %>%
  arrange(bird) %>%
  mutate(group = str_remove(group, "-"))

annot <- annot %>%
  separate(transcript_id, into = c(NA, "pt1", "pt2", "gene", "isoform")) %>%
  unite(transcript_id, pt1, pt2, gene, isoform) %>% 
  separate(gene_id, into = c(NA, "pt1", "pt2", "gene")) %>%
  unite(gene_id, pt1, pt2, gene)
```

```{r, warning = FALSE, message = FALSE}
bird <- as.factor(covars$bird)
sex <- as.factor(covars$sex)
age <- as.numeric(covars$age)
weight <- covars$wt_55
group <- recode(covars$group, C1 = "Ctl", C14 = "Ctl") %>% 
  as.factor()
pool <- as.factor(covars$Pool.Bursa)
ssgroup.virus.sac <- as.factor(covars$SSgroup.virus.sac)

covars.tib <- data.frame(bird,
                    sex,
                    age,
                    weight,
                    group,
                    pool,
                    ssgroup.virus.sac) %>% 
  as_tibble()
```

```{r, warning = FALSE, message = FALSE, results = 'hide'}
#Convert to DGEList object
dge <- DGEList(counts=cnt)
dge$genes <- annot[match(rownames(dge$counts), annot$gene_id),]

#CPM and log-CPM
cpm <- cpm(dge)
lcpm <- cpm(dge, log = TRUE)

#### Retain only genes expressed at >0.5 CPM in at least 25% of individuals
table(rowSums(dge$counts==0) == length(cnt[1,]))
keep.exprs <- rowSums(cpm>0.5) >= length(cnt[1,])/4
sum(keep.exprs)

dge <- dge[keep.exprs,, keep.lib.sizes=FALSE]
table(rowSums(dge$counts==0) == length(cnt[1,]))
dim(dge)

#TMM normalization
dge <- calcNormFactors(dge, method = "TMM")

#Assign groups in DGElist
dge$samples$group <- covars.tib$group

### Divide out I1 and I3
dge.I1 <- dge[,grep('\\bI1\\b', dge$samples$group)]
dge.I3 <- dge[,grep('I3', dge$samples$group)]

covars.I1 <- covars.tib %>% filter(group == "I1")
covars.I3 <- covars.tib %>% filter(group == "I3")
```

# Differential expression analysis
For this analysis, we conduct two sets of differential expression analyses- one for each LPAIV-infected group in early infection: I1 and I3.
```{r, warning = FALSE, message = FALSE,}
## Establish design matrix: I1
design.I1 = model.matrix(~ 0 +
                        factor(covars.I1$ssgroup.virus.sac) +
                        covars.I1$age +
                        covars.I1$sex + 
                        covars.I1$weight + 
                        covars.I1$pool)

colnames(design.I1) <- c("HIGH", "LOW", "MODERATE",
                         "age", "sexM", "weight", "pool2", "pool3")


contr.matrix.I1 = makeContrasts(
  LvM = LOW - MODERATE,
  MvH = MODERATE - HIGH,
  LvH = LOW - HIGH,
  levels = colnames(design.I1))

## Mean-variance trend plots
v.I1 <- voomWithQualityWeights(dge.I1, design.I1, plot = FALSE)

#Fitting models
vfit.I1 <- lmFit(v.I1, design.I1) 
vfit.I1 <- contrasts.fit(vfit.I1, contrasts = contr.matrix.I1)
tfit.I1 <- treat(vfit.I1, lfc = 1.0)

## Identify DE genes
dt.I1 <- decideTests(tfit.I1, p.value = 0.1, adjust.method = "fdr")
dt.tib.I1 <- as_tibble(dt.I1, rownames = NA) %>% 
  rownames_to_column("gene") %>% 
  mutate_at(vars(starts_with("L")), as.numeric) %>% 
  mutate_at(vars(starts_with("M")), as.numeric) %>% 
  rename(LvM.I1 = LvM,
         MvH.I1 = MvH,
         LvH.I1 = LvH)


## Establish design matrix: I3
design.I3 = model.matrix(~ 0 +
                        factor(covars.I3$ssgroup.virus.sac) +
                        covars.I3$age +
                        covars.I3$sex + 
                        covars.I3$weight + 
                        covars.I3$pool)

colnames(design.I3) <- c("HIGH", "LOW", "MODERATE",
                         "age", "sexM", "weight", "pool2", "pool3")


contr.matrix.I3 = makeContrasts(
  LvM = LOW - MODERATE,
  MvH = MODERATE - HIGH,
  LvH = LOW - HIGH,
  levels = colnames(design.I3))

## Mean-variance trend plots
v.I3 <- voomWithQualityWeights(dge.I3, design.I3, plot = FALSE)

#Fitting models
vfit.I3 <- lmFit(v.I3, design.I3) 
vfit.I3 <- contrasts.fit(vfit.I3, contrasts = contr.matrix.I3)
tfit.I3 <- treat(vfit.I3, lfc = 1.0)

## Identify DE genes
dt.I3 <- decideTests(tfit.I3, p.value = 0.1, adjust.method = "fdr")
dt.tib.I3 <- as_tibble(dt.I3, rownames = NA) %>% 
  rownames_to_column("gene") %>% 
  mutate_at(vars(starts_with("L")), as.numeric) %>% 
  mutate_at(vars(starts_with("M")), as.numeric) %>% 
  rename(LvM.I3 = LvM,
         MvH.I3 = MvH,
         LvH.I3 = LvH)

## Bring it all together
dt.tib <- left_join(dt.tib.I1, dt.tib.I3) %>% 
  mutate_at(vars(starts_with("L")), as.numeric) %>% 
  mutate_at(vars(starts_with("M")), as.numeric) %>% 
  filter_at(vars(LvM.I1:LvH.I3), any_vars(. != 0))
```

### Numbers of DE genes
```{r}
allResults <- rbind(
  as.data.frame(summary(dt.I1)),
  as.data.frame(summary(dt.I3))) %>% 
  filter(Var1 != "NotSig") %>% 
  rename("Direction" = Var1,
         "Comparison" = Var2,
         "N" = Freq)

kbl(allResults, booktabs = T, caption = "Count of DE genes. For a gene to be considered differentially expressed, we require a p-value of 0.1 with a false discovery rate correction and a log fold change difference of 1.0.") %>% 
  pack_rows("I1", 1, 6) %>%
  pack_rows("I3", 7, 12) %>%
  kable_styling(position ="center", latex_options = c("hold_position"))

```

\newpage 

### Volcano plot  
Volcano plot reporting -log10(p-values) as a function of log2(fold change) between the samples (logFC, x axis). Genes that are identified as significantly differentially expressed following a false discovery rate correction (q = 0.10) are shown in red  
```{r, warning = FALSE, message = FALSE, fig.height = 8, fig.width = 7, fig.align = "center"}
tmp1 <- topTreat(tfit.I1, coef = 1, n = Inf)
results1 <- mutate(tmp1, sig=ifelse(tmp1$adj.P.Val<0.1, "Sig", "Not Sig"))
p1 <- ggplot(results1, aes(logFC, -log10(P.Value))) +
  geom_point(aes(col = sig)) +
  scale_color_manual(values=c("black", "red")) +
  ggtitle("Low vs. Moderate shedding, Day 1") +
  ylab("-Log10(Adjusted p-value)") +
  xlab("Log(Fold count)") +
  theme_bw() +
  theme(legend.position = "none",
        plot.title = element_text(size = 9))


tmp2 <- topTreat(tfit.I1, coef = 2, n = Inf)
results2 <- mutate(tmp2, sig=ifelse(tmp2$adj.P.Val<0.1, "Sig", "Not Sig"))
p2 <- ggplot(results2, aes(logFC, -log10(P.Value))) +
  geom_point(aes(col = sig)) +
  scale_color_manual(values=c("black", "red")) +
  ggtitle("Moderate vs. High shedding, Day 1") +
  ylab("-Log10(Adjusted p-value)") +
  xlab("Log(Fold count)") +
  theme_bw() +
  theme(legend.position = "none",
        plot.title = element_text(size = 9))

tmp3 <- topTreat(tfit.I1, coef = 3, n = Inf)
results3 <- mutate(tmp3, sig=ifelse(tmp3$adj.P.Val<0.1, "Sig", "Not Sig"))
p3 <- ggplot(results3, aes(logFC, -log10(P.Value))) +
  geom_point(aes(col = sig)) +
  scale_color_manual(values=c("black", "red")) +
  ggtitle("Low vs. High shedding, Day 1") +
  ylab("-Log10(Adjusted p-value)") +
  xlab("Log(Fold count)") +
  theme_bw() +
  theme(legend.position = "none",
        plot.title = element_text(size = 9))

tmp4 <- topTreat(tfit.I3, coef = 1, n = Inf)
results4 <- mutate(tmp4, sig=ifelse(tmp4$adj.P.Val<0.1, "Sig", "Not Sig"))
p4 <- ggplot(results4, aes(logFC, -log10(P.Value))) +
  geom_point(aes(col = sig)) +
  scale_color_manual(values=c("black", "red")) +
  ggtitle("Low vs. Moderate shedding, Day 3") +
  ylab("-Log10(Adjusted p-value)") +
  xlab("Log(Fold count)") +
  theme_bw() +
  theme(legend.position = "none",
        plot.title = element_text(size = 9))

tmp5 <- topTreat(tfit.I3, coef = 2, n = Inf)
results5 <- mutate(tmp5, sig=ifelse(tmp5$adj.P.Val<0.1, "Sig", "Not Sig"))
p5 <- ggplot(results5, aes(logFC, -log10(P.Value))) +
  geom_point(aes(col = sig)) +
  scale_color_manual(values=c("black", "red")) +
  ggtitle("Moderate vs. High shedding, Day 3") +
  ylab("-Log10(Adjusted p-value)") +
  xlab("Log(Fold count)") +
  theme_bw() +
  theme(legend.position = "none",
        plot.title = element_text(size = 9))

tmp6 <- topTreat(tfit.I3, coef = 3, n = Inf)
results6 <- mutate(tmp6, sig=ifelse(tmp6$adj.P.Val<0.1, "Sig", "Not Sig"))
p6 <- ggplot(results6, aes(logFC, -log10(P.Value))) +
  geom_point(aes(col = sig)) +
  scale_color_manual(values=c("black", "red")) +
  ggtitle("Low vs. High shedding, Day 3") +
  ylab("-Log10(Adjusted p-value)") +
  xlab("Log(Fold count)") +
  theme_bw() +
  theme(legend.position = "none",
        plot.title = element_text(size = 9))

grid.arrange(p1, p2, p3, p4, p5, p6, nrow = 3)
```

\newpage 

```{r, eval = FALSE, warning = FALSE, message = FALSE, fig.height = 7, fig.width = 10, fig.align = "center"}
# Get the gene names for DE genes
lcpm2 <- as_tibble(dge.I1$counts) %>% 
  bind_cols(as_tibble(dge.I3$counts)) %>% 
  cpm(., log = TRUE)

newNames <- colnames(lcpm2) %>% 
  as_tibble() %>% 
  separate(value, into = c(NA, NA, NA, "sample", NA)) %>% 
  bind_cols(bind_rows(covars.I1, covars.I3)) %>% 
  unite("newNames", ssgroup.virus.sac, group, bird)

colOrder <- bind_cols(bind_rows(covars.I1, covars.I3)) %>% 
  mutate(ssgroup.virus.sac = fct_relevel(ssgroup.virus.sac, "LOW", "MODERATE", "HIGH")) %>%
  arrange(ssgroup.virus.sac, group, bird) %>% 
  unite("newNames", ssgroup.virus.sac, group, bird)

colnames(lcpm2) <- newNames$newNames
lcpm2 <- lcpm2[, colOrder$newNames]

annotList <- annot %>%
  select(
    gene_id,
    sprot_Top_BLASTX_hit
  ) %>%
  filter(gene_id %in% dt.tib$gene) %>% 
  separate(sprot_Top_BLASTX_hit, into = c("SwissProt_GeneName", NA, NA, NA, NA, NA, NA), "\\^") %>% 
  rename(gene = gene_id)

deGeneCounts <- lcpm2 %>% 
  as_tibble() %>% 
  mutate(gene = rownames(dge.I1$counts)) %>% 
  filter(gene %in% dt.tib$gene) %>% 
  left_join(annotList) %>% 
  unite("label", gene:SwissProt_GeneName, sep = " - ") %>% 
  distinct(label, .keep_all = TRUE) %>% 
  column_to_rownames("label") %>% 
  as.matrix()

## Set up palette
mypalette <- brewer.pal(11,"RdYlBu")
morecols <- colorRampPalette(mypalette)

hc <- hclust(as.dist(1-cor(t(deGeneCounts))))

# Plot the heatmap
heatmap.2(deGeneCounts,
          Colv = FALSE,
          Rowv = as.dendrogram(hc),
          col = rev(morecols(50)),
          trace = "none",
          colsep = c(10, 20),
          dendrogram = "row",
          density.info = "none",
          key = TRUE,
          #main = "Ileum - Gene level",
          margins = c(10, 20),
          scale ="row")
```

```{r message=FALSE, warning=FALSE}
tmp1a <- tmp1 %>% select(gene_id, adj.P.Val, logFC, sprot_Top_BLASTX_hit, Kegg) %>% separate(sprot_Top_BLASTX_hit, into = c("SwissProt_GeneName", NA, NA, NA, NA, "sprot2", NA), "\\^") %>%   separate(sprot2, sep = "=", into = c(NA, "SwissProt_GeneFunction")) %>%  separate(Kegg, into = c("KEGG_ID", "KO_ID"), sep = "\\`") %>%  mutate(comp = "LvM.I1")
tmp2a <- tmp2 %>% select(gene_id, adj.P.Val, logFC, sprot_Top_BLASTX_hit, Kegg) %>% separate(sprot_Top_BLASTX_hit, into = c("SwissProt_GeneName", NA, NA, NA, NA, "sprot2", NA), "\\^") %>%   separate(sprot2, sep = "=", into = c(NA, "SwissProt_GeneFunction")) %>%  separate(Kegg, into = c("KEGG_ID", "KO_ID"), sep = "\\`") %>% mutate(comp = "MvH.I1")
tmp3a <- tmp3 %>% select(gene_id, adj.P.Val, logFC, sprot_Top_BLASTX_hit, Kegg) %>% separate(sprot_Top_BLASTX_hit, into = c("SwissProt_GeneName", NA, NA, NA, NA, "sprot2", NA), "\\^") %>%   separate(sprot2, sep = "=", into = c(NA, "SwissProt_GeneFunction")) %>%  separate(Kegg, into = c("KEGG_ID", "KO_ID"), sep = "\\`") %>% mutate(comp = "LvH.I1")
tmp4a <- tmp4 %>% select(gene_id, adj.P.Val, logFC, sprot_Top_BLASTX_hit, Kegg) %>% separate(sprot_Top_BLASTX_hit, into = c("SwissProt_GeneName", NA, NA, NA, NA, "sprot2", NA), "\\^") %>%   separate(sprot2, sep = "=", into = c(NA, "SwissProt_GeneFunction")) %>%  separate(Kegg, into = c("KEGG_ID", "KO_ID"), sep = "\\`") %>% mutate(comp = "LvM.I3")
tmp5a <- tmp5 %>% select(gene_id, adj.P.Val, logFC, sprot_Top_BLASTX_hit, Kegg) %>% separate(sprot_Top_BLASTX_hit, into = c("SwissProt_GeneName", NA, NA, NA, NA, "sprot2", NA), "\\^") %>%   separate(sprot2, sep = "=", into = c(NA, "SwissProt_GeneFunction")) %>%  separate(Kegg, into = c("KEGG_ID", "KO_ID"), sep = "\\`") %>% mutate(comp = "MvH.I3")
tmp6a <- tmp6 %>% select(gene_id, adj.P.Val, logFC, sprot_Top_BLASTX_hit, Kegg) %>% separate(sprot_Top_BLASTX_hit, into = c("SwissProt_GeneName", NA, NA, NA, NA, "sprot2", NA), "\\^") %>%   separate(sprot2, sep = "=", into = c(NA, "SwissProt_GeneFunction")) %>%  separate(Kegg, into = c("KEGG_ID", "KO_ID"), sep = "\\`") %>% mutate(comp = "LvH.I3")

bind_rows(tmp1a, tmp2a, tmp3a, tmp4a, tmp5a, tmp6a) %>% 
  select(-SwissProt_GeneFunction, -KEGG_ID, -KO_ID) %>% 
  filter(gene_id %in% dt.tib$gene) %>% 
  mutate(logFoldCount = ifelse(adj.P.Val < 0.10, round(logFC, 2), "ns")) %>% 
  select(-adj.P.Val, -logFC) %>% 
  pivot_wider(names_from = comp, values_from = logFoldCount) %>% 
  kbl(booktabs = T, 
      longtable =T,
      caption ="Annotations for differentially expressed genes. ns denotes non-significant genes for each comparison and numerical values are the log(fold change) difference") %>%
  kable_styling(latex_options = c("repeat_header"))
```

```{r message=FALSE, warning=FALSE}
bind_rows(tmp1a, tmp2a, tmp3a, tmp4a, tmp5a, tmp6a) %>% 
  select(-comp, -adj.P.Val, -logFC) %>% 
  filter(gene_id %in% dt.tib$gene) %>% 
  distinct(.) %>% 
  select(-gene_id) %>% 
  filter(SwissProt_GeneName != ".") %>% 
  kbl(booktabs = TRUE, 
      longtable =T,
      caption = "Gene functions") %>%
  kable_styling(latex_options = c("repeat_header")) %>% 
  landscape() %>% 
  column_spec(2, width ="10em") %>% 
  column_spec(3, width ="15em")
```

```{r message=FALSE, warning=FALSE}
topGO.mappings <- readMappings("G:/Shared drives/RNA Seq Supershedder Project/BWTE DE manuscript/extData/GOdbGene.txt", sep = "\t", IDsep = ",")

goEnrich <- function(targetComp, ...) {
  DE.results <- dt.tib %>% filter(!!sym(targetComp) != 0)
  
  all.genes <- sort(unique(as.character(rownames(dge$counts))))
  int.genes <- DE.results$gene
  int.genes <- factor(as.integer(all.genes %in% int.genes))
  names(int.genes) = all.genes
  
  go.obj.BP <- new("topGOdata", ontology='BP',
              allGenes = int.genes,
              annot = annFUN.gene2GO,
              nodeSize = 5,
              gene2GO = topGO.mappings)

  go.obj.MF <- new("topGOdata", ontology='MF',
                 allGenes = int.genes,
                 annot = annFUN.gene2GO,
                 nodeSize = 5,
                 gene2GO = topGO.mappings)

  go.obj.CC <- new("topGOdata", ontology='CC',
                 allGenes = int.genes,
                 annot = annFUN.gene2GO,
                 nodeSize = 5,
                 gene2GO = topGO.mappings)
  
  results.BP <- runTest(go.obj.BP, algorithm = "elim", statistic = "fisher")

  results.tab.BP <- GenTable(object = go.obj.BP, 
                          elimFisher = results.BP, 
                          topNodes = 50) %>% 
    as_tibble() %>% 
    mutate(pVal = as.numeric(elimFisher)) %>% 
    filter(pVal < 0.01) %>% 
    mutate(Domain = "BP") %>% 
    mutate(Comparison = targetComp)
  
  results.MF <- runTest(go.obj.MF, algorithm = "elim", statistic = "fisher")
  
  results.tab.MF <- GenTable(object = go.obj.MF, 
                          elimFisher = results.MF, 
                          topNodes = 50) %>% 
    as_tibble() %>% 
    mutate(pVal = as.numeric(elimFisher)) %>% 
    filter(pVal < 0.01) %>% 
    mutate(Domain = "MF") %>% 
    mutate(Comparison = targetComp)
  
  results.CC <- runTest(go.obj.CC, algorithm = "elim", statistic = "fisher")
  
  results.tab.CC <- GenTable(object = go.obj.CC, 
                          elimFisher = results.CC, 
                          topNodes = 50) %>% 
    as_tibble() %>% 
    mutate(pVal = as.numeric(elimFisher)) %>% 
    filter(pVal < 0.01) %>% 
    mutate(Domain = "CC") %>% 
    mutate(Comparison = targetComp)
  
  rbind(results.tab.BP,
    results.tab.MF,
    results.tab.CC)
}

results <- list()

for(z in unique(names(dt.tib)[-1])) {
  if (dt.tib %>% 
      filter(!!sym(z) != 0) %>% 
      nrow(.) > 0)
  results[[length(results)+1]] <- goEnrich(z)
}

bind_rows(results, .id = "column_label") %>%
  dplyr::select(-column_label) %>% 
  arrange(GO.ID, Comparison) %>% 
  kbl(booktabs = T, 
      longtable = T,
      caption = "Gene Ontology enrichment analysis: GO terms with p-values <0.05 were determined as over-represented (enriched)") %>%
  kable_styling(latex_options =c("repeat_header")) %>% 
  landscape() 
```
